### ISSUE TITLE
execCommand starts shell without HOME set

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#7936]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)| @jml reported|
|-|-|
|Trac ID|trac#7936|
|Type|enhancement|
|Created|2015-06-11 14:38:45Z|
|Branch|https://github.com/twisted/twisted/tree/unix-conch-exec-command-home-7936-2|

`twisted.conch.ssh.SSHSessionForUnixConchUser.execCommand` launches a shell, but doesn't set `HOME` in the environment.

This is probably because it assumes that the `openShell` method has been called first. Calling `openShell` mutates the `environ` member variable which is used for spawning the shell in `execCommand`.

Because `HOME` is not set, many shell scripts (notably, virtualenvwrapper.sh) will fail with surprising errors.

<details><summary>Searchable metadata</summary>

```
trac-id__7936 7936
type__enhancement enhancement
reporter__jml jml
priority__normal normal
milestone__ 
branch__branches_unix_conch_exec_command_home_7936_2 branches/unix-conch-exec-command-home-7936-2
branch_author__ 
status__closed closed
resolution__fixed fixed
component__conch conch
keywords__ 
time__1434033525885420 1434033525885420
changetime__1435403320916401 1435403320916401
version__None None
owner__jml jml
cc__z3p
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_2"></a>|@jml commented|
|-|-|

(In [45028]) Branching to unix-conch-exec-command-home-7936.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_3"></a>|@jml commented|
|-|-|

(In [45113]) Branching to unix-conch-exec-command-home-7936-2.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_4"></a>|@jml commented|
|-|-|

The smallest possible tested fix for this specific problem that I could come up with. 

I'd like to reduce some of the duplication in the code, but am afraid of getting caught in a tarpit. I'd rather do lots of small, focused tickets.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_5"></a>|@hawkowl set owner to @jml|
|-|-|

Hi! Thanks for this, jml :)

Big things:

- Makes the Windows build break (https://buildbot.twistedmatrix.com/builders/windows7-64-py2.7/builds/4904/steps/iocp/logs/problems)
- TwistedChecker says ```W9012: 53,0: Expected 2 blank lines, found 1``` in twisted.conch.test.test_unix

Minor nitpicks:

- ```__future__``` imports would be nice on the new file
- in test_unix.py, please don't use the () style imports -- see https://github.com/twisted/twisted/compare/trunk...trialrunner-py3-5965-3#diff-004ff81390b7ca3b7a936cfcfad5f10dL12 for how they should look
- "Pretend to set the TCP_NODELAY" has a superflous "the" and should probably have "C{TCP_NODELAY}"
- Maybe add the link to the new tests at the top of the file? (the ```# -*- test-case-name:``` stuff)

Otherwise, looks good. Please fix the major issues, fix the nitpicks as you see fit, rerun the builders, and feel free to merge once the builders are green.

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_6"></a>|@jml commented|
|-|-|

(In [45115]) Address review comments

Refs [#7936](https://github.com/twisted/twisted/issues/7936)

* Gate tests on being able to import unix
* Use epytext correctly
* Use relative import syntax
* Don't use `()` for multi-line imports
* Add test-case-name file local variable
* Correct number of blank lines
* Drop superfluous 'the' from docstring

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_8"></a>|@jml commented|
|-|-|

I think I've made all the necessary changes. The twistedchecker builder fails, but I don't know how to read the failures.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_9"></a>|@hawkowl commented|
|-|-|

As mentioned in IRC, please replace the try-except method of importing unix with ```twisted.python.reflect.requireModule```, and then land.

Thanks again!

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_10"></a>|@jml commented|
|-|-|

(In [45116]) Use requireModule and explicitly check for None

Refs [#7936](https://github.com/twisted/twisted/issues/7936)

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_11"></a>|@jml commented|
|-|-|

(In [45117]) News file

Refs [#7936](https://github.com/twisted/twisted/issues/7936)

**Comment by twisted-trac** [comment]
|[<img alt="jml's avatar" src="https://avatars.githubusercontent.com/u/125674?s=50" width="50" height="50">](https://github.com/jml)<a name="note_12"></a>|@jml set status to `closed`|
|-|-|

(In [45118]) Merge unix-conch-exec-command-home-7936-2: Set HOME on execCommand

Author: jml
Reviewer: hawkowl
Fixes: [#7936](https://github.com/twisted/twisted/issues/7936)

Previously, Conch's Unix server would only set HOME when running shell
commands. Now, it also sets it when running commands without a shell.

Much of the time to find and fix this bug was provided by ClusterHQ.
